File: src/#508/solution.js
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

var findFrequentTreeSum = function (root) {
  let m = new Map();
  let res = [];
  traverse(root);
  let max = -1;
  m.forEach((count, sum) => {
    if (count >= max) {
      if (count > max) {
        max = count;
        res = [];
      }
      res.push(sum);
    }
  });

  return res;

  function traverse(root) {
    if (root === null) {
      return 0;
    }
    let total = root.val + traverse(root.left) + traverse(root.right);

    if (m.has(total)) {
      m.set(total, m.get(total) + 1);
    } else {
      m.set(total, 1);
    }
    return total;
  }
};

File: src/#508/solution.go
package solution

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func findFrequentTreeSum(root *TreeNode) []int {
	m := make(map[int]int)
	traverse(root, m)
	res := make([]int, 0)

	max := -1
	for sum, cnt := range m {
		if max <= cnt {
			if max < cnt {
				max = cnt
				res = res[:0]
			}
			res = append(res, sum)
		}
	}
	return res
}

func traverse(root *TreeNode, m map[int]int) int {
	if root == nil {
		return 0
	}

	leftTotal := traverse(root.Left, m)
	rightTotal := traverse(root.Right, m)

	total := root.Val + leftTotal + rightTotal

	m[total] += 1

	return total
}
